plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/DirectoryListingCacheKey.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.fs;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.plugin.hive.metastore.Table;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.util.List;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A cache key designed for use in {@link CachingDirectoryLister} and {@link TransactionScopeCachingDirectoryLister}
+ * that allows distinct cache entries to be created for both recursive, files-only listings and shallow listings
+ * (that also might contain directories) at the same {@link Path}, ie: {@link DirectoryLister#list(FileSystem, Table, Path)}
+ * and {@link DirectoryLister#listFilesRecursively(FileSystem, Table, Path)} results.
+ *
+ * It does this by pre-computing the hash code and uses the sign bit to indicate whether the key is a full recursive files
+ * listing or a shallow listing that could include directories. This then requires that the hashCode be consulted as part
+ * of {@link DirectoryListingCacheKey#equals(Object)}.
+ */
+final class DirectoryListingCacheKey
+{
+    private static final int RECURSIVE_FILES_ONLY_MASK = (1 << 31);
+    private final Path path;
+    private final int hashCode; // cached hashCode, embedding a bitmask value for whether the cache is a files-only, recursive listing
+
+    public DirectoryListingCacheKey(Path path, boolean recursiveFilesOnly)
+    {
+        this.path = requireNonNull(path, "path is null");
+        int hashCode = path.hashCode();
+        if (recursiveFilesOnly) {
+            // Set the sign bit to 1
+            hashCode = hashCode | RECURSIVE_FILES_ONLY_MASK;
```

Comment:
This feels line an overoptimization. I would probably just add a boolean field. With `4B` allocated for `hashCode` and potentially `~100B` on average allocated to store the `path` it doesn't feel like a `1B` boolean field will make any significant difference.

Reply:
Will store the `recursiveFilesOnly` boolean flag in it's own field instead.